core/domain/stats_services.py

```diff
@@ -352,76 +290,68 @@ def update_exp_issues_for_new_exp_version(
-        for p_idx, playthrough in enumerate(playthroughs):
-            if stats_models.CUSTOMIZATION_ARG_WHICH_IDENTIFIES_ISSUE[
-                    playthrough.issue_type] == 'state_names':
-                state_names = playthrough.issue_customization_args[
-                    'state_names']['value']
-                playthrough.issue_customization_args['state_names']['value'] = [
-                    new_state_name
-                    if state_name == old_state_name else state_name
-                    for state_name in state_names]
-            else:
-                playthrough.issue_customization_args['state_name']['value'] = (
-                    new_state_name)
-            for a_idx, action in enumerate(playthrough.actions):
-                if action.action_customization_args['state_name']['value'] == (
-                        old_state_name):
-                    playthroughs[p_idx].actions[
-                        a_idx].action_customization_args['state_name'][
-                            'value'] = new_state_name
-
-        all_playthrough_ids.extend(playthrough_ids)
-        all_playthroughs.extend(playthroughs)
-
-    update_playthroughs_multi(all_playthrough_ids, all_playthroughs)
+    deleted_state_names = exp_versions_diff.deleted_state_names
+    old_to_new_state_names = exp_versions_diff.old_to_new_state_names
+
+    playthrough_ids = list(itertools.chain.from_iterable(
+        i.playthrough_ids for i in exp_issues.unresolved_issues))
+    playthrough_models = (
+        stats_models.PlaythroughModel.get_multi(playthrough_ids))
+
+    for playthrough_model in playthrough_models:
+        playthrough = get_playthrough_from_model(playthrough_model)
+
+        if 'state_names' in playthrough.issue_customization_args:
+            state_names = (
+                playthrough.issue_customization_args['state_names']['value'])
+            playthrough.issue_customization_args['state_names']['value'] = [
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name] for state_name in state_names
+            ]
+
+        if 'state_name' in playthrough.issue_customization_args:
+            state_name = (
+                playthrough.issue_customization_args['state_name']['value'])
+            playthrough.issue_customization_args['state_name']['value'] = (
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name])
+
+        for action in playthrough.actions:
+            state_name = (
+                action.action_customization_args['state_name']['value'])
+            action.action_customization_args['state_name']['value'] = (
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name])
+
+        playthrough_model.populate(**playthrough.to_dict())
+
+    transaction_services.run_in_transaction(
```

Comment:
Can you explain in an inline comment why a transaction is needed here to provide context both to me & future folks who come across this code?

Reply:
Done.

Reply:
Please note: This PR (in particular this change) broke production. The transaction failed due to "BadRequestError: operating on too many entity groups in a single transaction." As a result, the "unresolved_issues" code after this was not executed, and some models are now missing their unresolved_issues field, which is leading to serious errors in prod.

@kevintab95 and I are working on trying to find a fix for this, but I wanted to point this out so that you're aware.

Reply:
Hmm, can we rollback these transaction changes in isolation?

Reply:
No. Models on production are now inconsistent because there isn't an easy way to figure out what has been affected and what hasn't. We are still not entirely sure how to deal with this.

Please ping me if you want to participate in the problem-solving.